api/v1alpha3/awsmachine_webhook.go

```diff
@@ -0,0 +1,55 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	"errors"
+	"reflect"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+func (r *AWSMachine) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha3-awsmachine,mutating=false,failurePolicy=fail,groups=infrastructure.cluster.x-k8s.io,resources=awsmachines,versions=v1alpha3,name=validation.awsmachine.infrastructure.x-k8s.io
+
+var _ webhook.Validator = &AWSMachine{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *AWSMachine) ValidateCreate() error {
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *AWSMachine) ValidateUpdate(old runtime.Object) error {
+	oldAWSMachine := old.(*AWSMachine)
+	if !reflect.DeepEqual(r.Spec, oldAWSMachine.Spec) {
+		return errors.New("awsMachineSpec is immutable")
```

Comment:
```suggestion
		return field.Forbidden(field.NewPath("spec"), "cannot be modified")
```

Reply:
I suspect we'll need to allow updates to at least a subset of the spec (controller setting `awsmachine.Spec.ProviderID` for example).

We could probably allow updates to the `AdditionalTags` and `AdditionalSecurityGroups` fields safely as well (since those should be updated on each reconcile).

It might also be helpful for people that are troubleshooting issues with initial instance creation to be able to modify fields like the AMI, ImageLookupOrg, InstanceType, IAMInstanceProfile, AvailabilityZone, Subnet, SSHKeyName, NetworkInterfaces, etc assuming the instance doesn't exist.

Reply:
Would it make sense to allow all updates as long as we haven't created an ec2 instance yet? We might need to worry about race conditions such as:

1. Simultaneously:
    1. Reconciler receives AWSMachine at ResourceVersion=1
    1. User quickly gets in some updates, now ResourceVersion=2
1. Reconciler creates ec2 instance using data from ResourceVersion=1 (maybe it's slow today)
1. Reconciler sees updated AWSMachine at ResourceVersion=2
    1. But it can't do anything at this point, so the data doesn't match what happened in ec2

Reply:
@ncdc as long as we don't necessarily rely on something like the providerID being set on the awsmachine.Spec, and fallback to doing a describe for the instance we should be able to avoid race conditions.

Reply:
@detiber I don't think that helps what I described above?

Reply:
Ah, I'm following now, I was thinking about a race condition resulting in orphaning. I suspect that yes it might be likely to be susceptible to the scenario that you provided.

Reply:
Which is better (or worse, depending on how you look at it)?

1. forcing the user to delete & recreate the AWSMachine to make changes (guarantees what we send to ec2 matches what's in the spec), or

2. allowing spec changes as long as we can't find the corresponding ec2 instance by tags (which might result in what we send to ec2 not matching the final state of the AWSMachine spec -- the scenario I described above)

Reply:
For this iteration, I'd keep it as simple as possible and disallow all updates, we can always iterate on the user experience later, when we have a better feeling how folks are using it

Reply:
That works for me

Reply:
I think my concern here is that I've seen multiple people recently hit issues with spinning up an instance and ask for help in channel. We've gotten them back on track by having them update the definition to fix the sshKey name, iam profile, etc. 

It's possible that this will not be an issue once we build out the control plane management, but it is something that people are hitting today and deletion/recreation and update of references to a new infra config would be a big regression for resolving those types of issues.

Reply:
I don't think you'd need to adjust the reference if you keep the name the same?

Reply:
We could move the logic that's now deleted in this PR into the webhook validation code.

Reply:
@ncdc our current implementation is also susceptible to this race condition you mentioned, right? so if we move the logic that's now deleted in this PR into the webhook validation code, it can also face this scenario?

Reply:
@tahsinrahman you are correct. I guess this comes back to my question from https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1218/files#r344289646: is it better to allow updates with the possibility of a race, or err on the side of caution and reject all updates?

Reply:
I'm on the side of disallowing all updates for now and see how it affects user experience. But lets come to an agreement on what to do! :)

Reply:
sgtm, seems like the general consensus is to disallow updates for now :)

Reply:
I think we have consensus!